core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -237,13 +227,12 @@ public synchronized IndexShard createShard(int sShardId, ShardRouting routing) {
         }
         final Settings indexSettings = settingsService.getSettings();
         final ShardId shardId = new ShardId(index, sShardId);
-        ShardLock lock = null;
+        final ShardLock lock = nodeEnv.shardLock(shardId, TimeUnit.SECONDS.toMillis(5));
         boolean success = false;
         Store store = null;
         IndexShard indexShard = null;
+        eventListener.beforeIndexShardCreated(shardId, indexSettings);
```

Comment:
can we move this back into the try? I'm worried that exceptions wouldn't release the shard lock  .